### PR TITLE
Fix Csharp Cmake

### PR DESCRIPTION
**Pull request by runette**
## What does this PR do?
1. Updates the default Dotnet SDK level from the "out of support" net5 to net6. No one should be using the default in production - they should explicitly set the SDK they want - but this will avoid "out of support" messages appearing if they do.

2. Adds a check that the requested Dotnet SDK is available on the machine and fails more gracefully if it is not.

## What are related issues/pull requests?

None

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Updated Python API documentation (swig/include/python/docs/)
 - [x] Review
 - [x] Adjust for comments
 - [x] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS: All
* Compiler: Dotnet / Cmake

### CODE REVIEW

**Comment by rouault** [comment]
@runette is that appropriate for backport in the stable branch ?

File: cmake/modules/thirdparty/FindDotnet.cmake [hunk 1]
```diff
@@ -137,6 +138,16 @@ EXECUTE_PROCESS(
     OUTPUT_STRIP_TRAILING_WHITESPACE
 )
 
+EXECUTE_PROCESS(
+    COMMAND ${DOTNET_EXE} --list-sdks
```
Comment (dg0yt) [inline cmake/modules/thirdparty/FindDotnet.cmake:142]: `${DOTNET_EXE}` should be in double quotes, to protect embedded space. You know `C:\Program Files...`.

This is something to keep in mind more generally for single values in CMake (but not always for lists of values). e.g. `DOTNET_SDKS` below. I'm aware that this insufficiently implemented in this file and in many other cmake files here.

File: swig/csharp/CMakeLists.txt [hunk 2]
```diff
@@ -16,15 +16,26 @@ set(GDAL_CSHARP_INSTALL_DIR
...
+    "net6.0"
     CACHE STRING ".NET version to be used for the sample Applications")
 set(CSHARP_CONFIG
     "RELEASE"
     CACHE STRING "Config to be used to compile the C# artifacts > RELEASE|CONFIG")
 
+string(REGEX MATCH "[0-9]+\.*[0-9]+" CSHARP_SDK ${CSHARP_LIBRARY_VERSION} )
+message(STATUS "Requested DotNet SDK Level: " ${CSHARP_SDK} )
+
+if ( DOTNET_SDKS MATCHES ".*${CSHARP_SDK}.*" )
```
Comment (dg0yt) [inline swig/csharp/CMakeLists.txt:31]: The leading and trailing `.*`  are pointless. And you must be sure that `${CSHARP_SDK}` contains no special characters (but even `.` is special). For an exact match, there is only `string(FIND ...)` for strings (or `IN_LISTS` for lists).

**Review by dg0yt — COMMENTED** [review]
Just some stylistic comments.

File: cmake/modules/thirdparty/FindDotnet.cmake (hunk 1 above)
Comment (runette) [reply cmake/modules/thirdparty/FindDotnet.cmake]: Good catch. Fixed in the latest update.

Interestingly, the original code that was written by Microsoft had the same bug ... :)

File: swig/csharp/CMakeLists.txt (hunk 2 above)
Comment (runette) [reply swig/csharp/CMakeLists.txt]: Your comment made me realise that while it was working, it was not actually checking for incorrect SDK specifications and led to rewrite it in a more robust way

**Comment by runette** [comment]
@rouault - I am not sure.

It would probably make sense

File: swig/csharp/CMakeLists.txt [hunk 3]
```diff
@@ -16,15 +16,27 @@ set(GDAL_CSHARP_INSTALL_DIR
...
-    "net5.0"
+    "net6.0"
     CACHE STRING ".NET version to be used for the sample Applications")
 set(CSHARP_CONFIG
     "RELEASE"
     CACHE STRING "Config to be used to compile the C# artifacts > RELEASE|CONFIG")
 
+message(STATUS "Available SDKS: " "${DOTNET_SDKS}")
+
+string(REGEX MATCH "[0-9\.]+" CSHARP_SDK "${CSHARP_LIBRARY_VERSION}" )
```
[the remark this answers is not in the document]
Comment (runette) [reply swig/csharp/CMakeLists.txt:30]: That comment has opened a few more issues. Will push a commit later

[the remark this answers is not in the document]
Comment (runette) [reply swig/csharp/CMakeLists.txt]: OK

- I was being a bit lazy with internal variables. That is fixed.
- The CSHARP_SDK parameter was not being used - which is not ideal. It was also doing the same thing as other parameters.
- I have removed the CSHARP_SDK parameter
- Fixed up the Mono compile routines to use the CSHARP_LIBRARY_VERSION and CSHARP_APPLICATION_VERSION parameters correctly

**Comment by runette** [comment]
@rouault - not sure what to do about the failed test - it appears to be some sort of time out and nothing to do with any changes I have made

**Comment by rouault** [comment]
> not sure what to do about the failed test - it appears to be some sort of time out and nothing to do with any changes I have made

yes, it is unrelated. I've restarted that build

**Comment by rouault** [comment]
The backport to `release/3.6` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit ae40a45d48f28273ca4403773386192f4effebd8 is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
[backport-6843-to-release/3.6 b2ac3c9ca2] fix error messages when incorrect SDK
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 17:30:34 2022 +0000
 2 files changed, 22 insertions(+), 4 deletions(-)
[backport-6843-to-release/3.6 29599bc84d] Update CMakeLists.txt
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 20:04:21 2022 +0000
 1 file changed, 4 insertions(+), 4 deletions(-)
[backport-6843-to-release/3.6 b634fcfd5b] debug
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 20:22:55 2022 +0000
 2 files changed, 7 insertions(+), 3 deletions(-)
[backport-6843-to-release/3.6 964254aedc] regex fix
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 20:45:49 2022 +0000
 1 file changed, 2 insertions(+), 2 deletions(-)
[backport-6843-to-release/3.6 10be45c18e] fix regex
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 20:56:17 2022 +0000
 2 files changed, 2 insertions(+), 2 deletions(-)
[backport-6843-to-release/3.6 2246cff0b4] regex fix
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 21:15:33 2022 +0000
 2 files changed, 7 insertions(+), 5 deletions(-)
[backport-6843-to-release/3.6 dc9c14c139] regex fix
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 21:31:26 2022 +0000
 1 file changed, 1 insertion(+), 1 deletion(-)
[backport-6843-to-release/3.6 d9963e081e] regex fixes
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 21:33:45 2022 +0000
 1 file changed, 1 insertion(+), 1 deletion(-)
[backport-6843-to-release/3.6 aad5978444] regex fixes
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 21:46:15 2022 +0000
 1 file changed, 2 insertions(+), 2 deletions(-)
[backport-6843-to-release/3.6 467b8ddde5] regex fixes
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 22:04:21 2022 +0000
 1 file changed, 2 insertions(+), 1 deletion(-)
[backport-6843-to-release/3.6 65a6c91e4a] remove debug messages
 Author: Paul <runette@gmail.com>
 Date: Fri Dec 2 22:19:47 2022 +0000
 2 files changed, 1 insertion(+), 4 deletions(-)
[backport-6843-to-release/3.6 aa44c80520] improve regexs
 Author: Paul <runette@gmail.com>
 Date: Sat Dec 3 09:46:27 2022 +0000
 2 files changed, 16 insertions(+), 8 deletions(-)
[backport-6843-to-release/3.6 0c1a61f567] regex fix
 Author: Paul <runette@gmail.com>
 Date: Sat Dec 3 10:01:25 2022 +0000
 2 files changed, 1 insertion(+), 4 deletions(-)
[backport-6843-to-release/3.6 aa3a86059b] fix typos
 Author: Paul <runette@gmail.com>
 Date: Sat Dec 3 10:11:42 2022 +0000
 1 file changed, 3 insertions(+), 2 deletions(-)
[backport-6843-to-release/3.6 a635e970cf] Update FindDotnet.cmake
 Author: Paul <runette@gmail.com>
 Date: Sat Dec 3 10:23:59 2022 +0000
 1 file changed, 4 insertions(+), 4 deletions(-)

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-release/3.6 release/3.6
# Navigate to the new working tree
cd .worktrees/backport-release/3.6
# Create a new branch
git switch --create backport-6843-to-release/3.6
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 5dca19f5f7d2e13d2be48f7989411ead11b22251,1697d4c40760ccb37ab7190d6e16a8ca27defe41,5a9c868b104dae3309381f66d124b5f888bb284a,1b929ba9262265ab5e2b76df64f0e807eb9079cc,83a126413bfcd645496a5e53b16bbe60a60cf725,a515c3f4835f76e41f3c7c7829da29a077786a83,19487af54143b152fdcccc2ac7ff3d4f3b68b410,5789c9a00f5ff53cdc097cc8ad4710c721e73e28,a8c5881345efb568503272784537eeb74c5cadfe,149dfee142e663352303caf8edb6047012781f93,2e950a9ca64d96c9aa019dec65e069199a86a8b8,a95d7af23cf8ece293edfb1d0f712a4596f04aa3,c4f28a29102f3b0de7553c8061bb0d70837aa5e6,649b363e83cefa02ad472d974a0ec433269978d2,4abeaf17cdc4896108c69b5eeb1da6cfc6e3c576,ae40a45d48f28273ca4403773386192f4effebd8,82c1953c102b541dc090ed1f84b4a610ae193670,5715451e047adc80ee4ba14e079270aee8e3db5f,f23d845419244793602a9855aa4235b9eaa8197a,de985545814bc0ad7b3dd911f9231b7ff1dd412b,2df5af161f7f02e3c1dd71f1333e86046d138212,60e45f21c217bac629160ade49d950fd0c89bbcc,71271652e3a188f9baebdb282824eb493761561b,43f3805eb8dcee1570b83fd6fa2eb8bdaa3a3ac7,bb818d3830f784ba79072e0adee03a4b73c676fb,27f07cef0b38d31e7e424765d19d6083d065bdfd,5d0db620af85ba1c084ef87d6d4abbbd01b7132a,5d3e33d46a130eb7366494dc1ed52ad3188a52c1,4d0690f0450d31f19eac31db234529f6a5f69692,d15c0a8fc8412351679a4ebe7b80e32b5f08792e
# Push it to GitHub
git push --set-upstream origin backport-6843-to-release/3.6
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-release/3.6
```
Then, create a pull request where the `base` branch is `release/3.6` and the `compare`/`head` branch is `backport-6843-to-release/3.6`.

**Comment by rouault** [comment]
manually cherry-picked in 3.6 per 5d9588b112be3aabac7278150fd6fe9ccbb1d598

**Pull request merged**